cascaritaco/cascarita
# Feature: Stripe webhook architecture

### Description 

**Breaking Change**
The webhook is set before the body-parser and any middleware in order to use [Strip Signature verification ](https://docs.stripe.com/webhooks#verify-webhook-signatures-with-official-libraries) library.  

**Handling Stripe Webhooks**

- In `app.js` we initiate the route and the actual logic can be found in the `controllers/stripeResources/connectedAccount.controller.js` 
- The webhook events will be handled by the `connectedAccount.controller.js` this file is meant to handle webhook events dealing with the connected accounts (a stripe account that is linked to Cascarita) . For now we only handling the updating of a connected account, reference this doc on tis type of event [`account.update` ](https://docs.stripe.com/api/events/types#event_types-account.updated) . We can make other files to handle other type of events based on resource (ie. `payment_method `, `customer` [etc](https://docs.stripe.com/api/events/list)) 
- wrote a migration updating UserStripeAccounts table and model.  Need to track more fields to know a connected account's status with stripe. 

### Checklist

- [x] I have tested the changes locally by running `npm run test`
- [ ] I have autoformatted the code by running `npm run eslint`
- [ ] I have added test cases (if applicable)

### Testing instructions

0. ensure '.env' has all the stripe values ! Contact me for `.env` values you all need to run stripe
1. check into branch and run migrations; cd to root of project directory and run the command:
- `npx sequelize-cli db:migrate`

2. run this seeder file to get a stripe account to test the webhook; cd to root of project directory and run the command:
- `npx sequelize db:seed --seed 20240902193649-demo-stripe-user`
- now you can log in, password is on the seed file

3. Open a terminal and run the following commands: 
- `npm run dev` to start running cascarita 
- in another terminal tab, run this command to turn on the webhook (this is for localhost testing purposes, for prod we will run run a similar thing but using stripe dashboard) 
a) install stripe cli : `brew install stripe/stripe-cli/stripe`
b)`stripe login --interactive`
c) it will ask for the api kay and that can be found in `env.STRIPE_TEST_API_KEY` 
d)`stripe listen --forward-to localhost:3001/api/webhook/stripe/connect`

4. trigger webhook:
- do it via postman, using stripe requires authentication codes and they are linked to my phone but we can trigger an account update using postman and auth tokens are already loaded in this post request.
- [post man request](https://solar-meadow-3701.postman.co/workspace/test-local-node-db~84fda4e7-bb80-462d-ae9f-6b752c287229/request/17973354-ff467dfe-e10f-48f6-8fc2-cbcb1a9e22a2?tab=body)
- here you can alter the metadata which will trigger the webhook.

5. Terminal for webhook activity: 
- keep the terminal tab open from step 3 where you ran the `stripe listen` command 
- when making updates via postman, you will see the all the events that action triggered. So far this pr only handles updates to the connect account, denoted by event type of `account.updated` for the rest of the events we return a `403`. As we continue to build cascarita we will handle more event types.
